### ISSUE TITLE
[Feature] Containerization for running via Docker

### ISSUE THREAD

**Issue by blobberbun** [Affirmatech/MeshSense#28]
It would be great to have automated docker builds of the most recent releases that would push to Dockerhub for easy deployment and updates

I believe this would restrict management to WiFi as Bluetooth pass through would be complex if not impossible

**Comment by Soltares** [comment]
Thanks for creating an issue for this.  I certainly appreciate Docker for large applications with arms and legs of dependencies.  Part of the design decisions for MeshSense, however, was to go with Electron to be very straightforward for deployment and updates.  Windows/Linux/Mac deployments are currently a single artifact that can execute and run with updates handled within the application.

I'm trying to rationalize the benefits of introducing docker as it almost seems like increased complexity.  For the official builds we also sign the artifacts using an in-house Affirmatech certificate and I'm not sure how that would play in the automation.

The only data files explicitly saved by the application live in `$HOME/.local/share/meshsense` (on Linux) in order to save a few state related information.  Otherwise the system shouldn't be polluted by anything outside of Electron standard operating procedures.

Not trying to dismiss Docker -- only trying to weigh the benefits vs time investment. :+1:

**Comment by corrosive4354** [comment]
maybe u can try [servbay](https://servbay.com),better than docker.

**Comment by PrintsLeo** [comment]
I'd also like to request this as an official feature.  For me, the benefit of having a docker container is that I don't need to have it running on a PC somewhere and it can run on my home server.  I can easily accomplish this by adding a VM and running it on the server.  However, a VM will use a lot more of the server's resources as it has to run an additional OS and the resources are dedicated to the VM.  

I do appreciate all your hard work in creating and supporting MeshSense!  It's an awesome tool!

**Comment by JavierLuna** [comment]
I'm in the same boat as the rest, I'd love to run headless meshsense within my home server where I already use docker heavily.

I think this docker feature request has less to do with polluting our environment and more with being able to have a virtualized headless instance of Meshsense using tooling that is commonly used in the homelab community.

The somewhat ideal setup for me would be to run Meshsense in headless mode in my home server, talking to a meshstastic node that I'd leave at home. I'd also front Meshsense with a reverse proxy to get that SSL goodness.

I would either access Meshsense from my local network or from outside by using a VPN that I already have running, but never from a desktop environment.

So I would see the docker container geared towards a different user experience (i.e 100% headless instead of desktop), and most importantly targeting a different user base (i.e homelabbers).

I think for me the lack of a docker image is not something that blocks me from using Meshsense, but it certainly makes it harder for me to install and maintain because it is a binary and I cannot use my "docker tools/runbooks" with it.

I'll try to containerize it on my fork today if I have time and ping you. No expectations of getting it merged here, just an opportunity to brush up my docker-fu 😄

**Comment by PrintsLeo** [comment]
> I'm in the same boat as the rest, I'd love to run headless meshsense within my home server where I already use docker heavily.
> 
> I think this docker feature request has less to do with polluting our environment and more with being able to have a virtualized headless instance of Meshsense using tooling that is commonly used in the homelab community.
> 
> The somewhat ideal setup for me would be to run Meshsense in headless mode in my home server, talking to a meshstastic node that I'd leave at home. I'd also front Meshsense with a reverse proxy to get that SSL goodness.
> 
> I would either access Meshsense from my local network or from outside by using a VPN that I already have running, but never from a desktop environment.
> 
> So I would see the docker container geared towards a different user experience (i.e 100% headless instead of desktop), and most importantly targeting a different user base (i.e homelabbers).
> 
> I think for me the lack of a docker image is not something that blocks me from using Meshsense, but it certainly makes it harder for me to install and maintain because it is a binary and I cannot use my "docker tools/runbooks" with it.
> 
> I'll try to containerize it on my fork today if I have time and ping you. No expectations of getting it merged here, just an opportunity to brush up my docker-fu 😄

There's already a docker image somebody else created, but I was unable to get it to work correctly.  When it starts the API service, it looks for a BT controller, which my server does not have, then crashes.  The owner of the docker did respond to one of my requests and was able to fix it, but the BT issue is part of MeshSense itself and not the docker.

https://hub.docker.com/r/y0mg/meshsense

**Comment by Soltares** [comment]
Thank you all for taking the time to provide feedback on this.  Others have also written in via e-mail and I am in agreement that there is a need for containerization for the purposes of running on an appliance or other integrated setup.

The next steps I need to address before tackling this are:
- How to handle #8 automatic updates in the container environment.  Is it the responsibility of the container owner to perform the updates?
- An elegant way to disable bluetooth functionality either via environment variable or a build with it removed as explored here: https://github.com/Affirmatech/MeshSense/issues/22#issuecomment-2445397087

**Comment by JavierLuna** [comment]
Thank you for listening, and for a great product 😄 

Just to give my two cents on the next steps you mentioned:

> How to handle automatic updates in the container environment. Is it the responsibility of the container owner to perform the updates?

I personally like to manually update the docker images, that way I don't get surprised by a breaking change and I can always pin the tag to a specific version if a new one is not working for me.

Having said that I know that other projects like to include an updater within the container. I usually disable the option, but I guess other folks might like it. I would consider it as a nice to have rather than something critical though.

> An elegant way to disable bluetooth functionality either via environment variable or a build with it removed as explored here

I haven't tried building the bluetooth library in arm (I have a pi and an M1 mac, so I can test in different devices too if you need me to), but I see https://github.com/thegecko/webbluetooth/issues/190 had success with building it so It should be possible to add bluetooth to an arm image?

Although I'm not sure how bluetooth works with a docker container 🤔 not sure if the host needs to pass the bluetooth device to the container.

**Comment by youroldmangaming** [comment]
Hi All 
It took a while but got it going

docker run --rm -v ./data:/root/.local/share/meshsense/ -v ./dbus/:/var/run/dbus/system_bus_socket:ro -p 5920:5920 y0mg/meshsense:amd64

AMD64 will get the RaspberryPi Version working soon.  I also have a fix for BT to include soon.


The Docker run works.

https://github.com/youroldmangaming/MeshSense

**Comment by PrintsLeo** [comment]
Thanks for your hard work as well!  For some reason, when I do an update/re-install, I have to manually install one of the dependencies.  I see that it gets installed, but I get the error below.  A manual install fixes it.  Not a big deal, but thought you might want to know.

Error: libdbus-1.so.3: cannot open shared object file: No such file or directory

Once I reinstall that package, I get the following error:

[State] bluetoothDeviceList init []
[bluetooth] Begin Scanning
Disconnecting from device
node:events:485
throw er; // Unhandled 'error' event

^

Error: connect ECONNREFUSED /var/run/dbus/system_bus_socket

at PipeConnectWrap.afterConnect [as oncomplete] (node:net:1615:16)
Emitted 'error' event on MessageBus instance at:

at EventEmitter.<anonymous> (/app/MeshSense/api/node_modules/dbus-next/lib/bus.js:159:12)
at EventEmitter.emit (node:events:507:28)
at Socket.<anonymous> (/app/MeshSense/api/node_modules/dbus-next/lib/connection.js:102:10)
at Socket.emit (node:events:507:28)
at emitErrorNT (node:internal/streams/destroy:170:8)
at emitErrorCloseNT (node:internal/streams/destroy:129:3)
at processTicksAndRejections (node:internal/process/task_queues:90:21) {
errno: -111,
code: 'ECONNREFUSED',
syscall: 'connect',
address: '/var/run/dbus/system_bus_socket'
}

Node.js v23.1.0
[31m[nodemon] app crashed - waiting for file changes before starting...[39m

**Comment by youroldmangaming** [comment]
OK this is running now on the Raspberry Pi 5

docker run --rm -v ./data:/root/.local/share/meshsense/ -v ./dbus/:/var/run/dbus/system_bus_socket:ro -p 5920:5920 y0mg/meshsense:arm64   Have remediated the BT, I think, as well.

**Comment by PrintsLeo** [comment]
Everything is working great in a Docker Container!

**Comment by jbouse** [comment]
> Thank you all for taking the time to provide feedback on this. Others have also written in via e-mail and I am in agreement that there is a need for containerization for the purposes of running on an appliance or other integrated setup.
> 
> The next steps I need to address before tackling this are:
> 
> * How to handle [Headless Mode #8](https://github.com/Affirmatech/MeshSense/issues/8) automatic updates in the container environment.  Is it the responsibility of the container owner to perform the updates?
> * An elegant way to disable bluetooth functionality either via environment variable or a build with it removed as explored here: [Support for arm64 Raspberry Pi #22 (comment)](https://github.com/Affirmatech/MeshSense/issues/22#issuecomment-2445397087)

As for the first item of updating the container, I would suggest potentially looking at https://github.com/containrrr/watchtower, which is what the Bluesky PDS deployment into Docker uses to update the containers.

**Comment by jbouse** [comment]
> OK this is running now on the Raspberry Pi 5
> 
> docker run --rm -v ./data:/root/.local/share/meshsense/ -v ./dbus/:/var/run/dbus/system_bus_socket:ro -p 5920:5920 y0mg/meshsense:arm64 Have remediated the BT, I think, as well.

So, if I'm following, your repo is just building the Docker image by cloning the upstream repo and applying your BT patch. It isn't a fork of this upstream repo?

**Comment by youroldmangaming** [comment]
Hi yes I did not change any of the MeshSense code. If you look at the Dockerfile I am pulling the git repo to build the image. https://github.com/youroldmangaming/MeshSense/blob/main/Dockerfile

**Comment by joeldhenry** [comment]
Also waiting on this.
